hashicorp/consul-k8s
# Add metrics merging and Prometheus annotations to connect-injected Pods

Adds metrics merging feature to consul-k8s.

Changes proposed in this PR:
- inject-connect takes default metrics flags
- supports annotations for overriding the default flags
- inject-connect passes metrics flags to connect-inject handler
- handler now has functions to grab configuration from the default flags or the overridden values from the annotations
- from the handler:
  - init container adds envoy_prometheus_bind_addr and flags to consul connect envoy
  - consul-sidecar is given the appropriate flags to run the merged metrics server
  - added prometheus annotations to pod to scrape from

How I've tested this PR:
- consul-helm (`test-metrics` branch)
- consul-k8s (`merged-metrics` branch)
- consul (`metrics-merging` branch)
- Build the consul and consul-k8s images or use:
  - `gcr.io/nitya-293720/consul-k8s-dev:metrics28`
  - `gcr.io/nitya-293720/consul-dev:metrics10`
  - Set those in values.yaml, make sure connectInject is enabled
- Deploy prometheus: `helm install prom prometheus-community/prometheus`
- Helm install consul
- `k apply -f example-app.yaml`
example-app.yaml: 
```
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app.kubernetes.io/name: prometheus-example-app
  name: prometheus-example-app
spec:
  replicas: 1
  selector:
    matchLabels:
      app.kubernetes.io/name: prometheus-example-app
  template:
    metadata:
      annotations:
        'consul.hashicorp.com/connect-inject': 'true'
      labels:
        app.kubernetes.io/name: prometheus-example-app
    spec:
      containers:
      - name: prometheus-example-app
        image: quay.io/brancz/prometheus-example-app:v0.3.0
        ports:
        - name: web
          containerPort: 8080
```
- Port forward to prometheus
  - `export POD_NAME=$(kubectl get pods --namespace default -l "app=prometheus,component=server" -o jsonpath="{.items[0].metadata.name}")`
  - `kubectl --namespace default port-forward $POD_NAME 9090`
- go to localhost:9090 and look at targets, you should see prometheus-example-app under kubernetes pods
- try graphing the "version" metric (an app metric) and an envoy metric

How I expect reviewers to test this PR:
- Give the steps above a shot
- Code review


Checklist:
